Suggestion: Create ansible-galaxy collection

dunnevan opened the pull request:
```
- Create galaxy.yml
- Add requirements.yml to show we have no external dependencies (to run the roles)
- Move custom_modules role to standard ansible layout (plugins)
```

@rajan-mis @whowutwut @mamuthiah 

I'm not sure why there are two duplicate PR's #103 and #104, but this change relates to them.  It would be nice to be able to consume the install-infra roles externally in other external playbooks and roles.  Part of this is conforming to ansible's standard practices for imports, which is ansible-galaxy collections.

I imagine this PR needs more work.  I did not add the playbooks to the collection, so it has the hacky symlinks for the modules.  Down the road it might be good to move the playbooks into a playbooks/ directory as dictated by the collection directory structure.

whowutwut commented:
Can this PR be modified to go against the `dev` branch without much pain?  it seems like it should be able to......  

The goal is that in the near future is that `dev` branch would merge to `master` and we can get back to a git flow process.  Otherwise we'll need to duplicate the PR to each branch.  Another option is we can review this , then after it's (squash) merged to `master` we can cherry-pick the merge commit to dev?  I still like just putting this into  dev though...

rajan-mis commented:
@dunnevan Thanks Evan.  Currently our master and dev branch is not in synced , so i always have to create 2 pull request , one for dev and one for master that's the reason ,there are two  PR's #103 and #104.

I will also try to validate this galaxy based directory collection and then will first try to merge into the dev branch . 

Can you please let me know , how  can we import this in the role . Thank you.

whowutwut commented:
Right now, since this is not being pushed to a Ansible Galaxy server, you would need to build it and it would generate a tar file: 

```
$ ansible-galaxy collection build
Created collection for ibm_spectrum_scale.install_infra at ....ibm-spectrum-scale-install-infra/ibm_spectrum_scale-install_infra-v0.0.0.tar.gz
```

Then you can move this tar file to where you plan to consume it and [install](https://docs.ansible.com/ansible/latest/user_guide/collections_using.html#installing-collections-with-ansible-galaxy):
```
ansible-galaxy collection install /path/to/ibm_spectrum_scale-install_infra-v0.0.0.tar.gz
```

If it eventually gets pushed somewhere public, 

Then you can call it multiple ways: [using-collections-in-playbook](https://docs.ansible.com/ansible/latest/user_guide/collections_using.html#using-collections-in-a-playbook), one example: 

```
- hosts: scale_cluster
  collections:
    - ibm_spectrum_scale.install_infra
  roles:
    - core/cluster
```

dunnevan commented:
> I will also try to validate this galaxy based directory collection and then will first try to merge into the dev branch .

I cherry-picked my branch onto dev, and switched this PR to the dev branch.

> Can you please let me know , how can we import this in the role . Thank you.

Collections are meant to be _installed_ and consumed by others.  You can use the collections keyword as @vhu showed.  Roles within a collection automatically have the plugins/modules and other roles in scope (imported).  However, the gotcha is - the collection has to be installed.

It seems ansible collection playbooks are not yet well developed, so I'm not sure how to run those directly from an installed collection.

Since, currently this project runs the playbook directly (_clone repo -> ansible-playbook playbook.yml_), that would mean benefit of the collection namespace (and imports) aren't there (since the collection would have to be installed first).

So, for the direct consumption I added the symlinks library/ and module_utils/ that is in the same directory as the existing playbook.  This is the mechanism for module imports at a playbook level.

Externally, when the collection is installed though - the modules and roles are imported properly with the collection mechanism.  This is how an external playbook would reference the collection
```
- hosts: scale_cluster
  gather_facts: no
  collections:
    - ibm_spectrum_scale.install_infra
  roles:
    - core/cluster
  vars:
    scale_version: 5.0.4.0
```

dunnevan commented:
I wanted to add a comment to be clear - this is just a suggestion that I think would be beneficial.  If this doesn't fit in the project, then it is fine to put this aside or close it.

whowutwut commented on galaxy.yml:
```diff
@@ -0,0 +1,22 @@
+namespace: ibm_spectrum_scale
+name: install_infra
+version: v0.0.0
```
Any idea how this version can be non static.... meaning, could this piggy back on the git tags/releases?

dunnevan replied:
I don't think it is possible, the closest you can get is some pre-submit hooks - but that only applies if you are making the tag on the command line I believe.

whowutwut commented:
@mamuthiah From our conversation this morning, I took the latest dev branch and cherry-picked these commits... then did a simple cluster deploy... and then set 1 node in my hosts file to `state=absent` , and ran the playbook again. 

With my limited knowledge here, it did remove the node from the cluster (no longer in mmlscluster) and this testing branch is using the modules defined in this path:  

```
[root@collection-test-x-master ibm-spectrum-scale-install-infra]# git grep "ibm_ss_node"
plugins/modules/ibm_ss_node.py:module: ibm_ss_node
plugins/modules/ibm_ss_node.py:  ibm_ss_node:
plugins/modules/ibm_ss_node.py:  ibm_ss_node:
plugins/modules/ibm_ss_node.py:  ibm_ss_node:
roles/core/cluster/tasks/removenode.yml:      ibm_ss_node:
```

I think this was your concern?

whowutwut commented:
FYI, some background on 351ccbbe0b01874d387b40458d5e36d95a3441a9, I tried to run these modules and calling in a simple playbook outside of the git repo, bu hit Errors like:

* `ERROR! couldn't resolve module/action 'ibm_ss_cluster'.`
* `ERROR! the role 'core/cluster' was not found in ... <bunch of ansible paths> `

In order for the collections to be located, we have to switch the way we import the modules since they are now part of collections:  See https://docs.ansible.com/ansible/latest/dev_guide/developing_collections.html#module-utils for more info, but essentially: 
> When coding with module_utils in a collection, the Python import statement needs to take into account the FQCN along with the ansible_collections convention. The resulting Python import will look like from ansible_collections.{namespace}.{collection}.plugins.module_utils.{util} import {something}

whowutwut commented:
I re-tested running the playbook from the git clone of this repo... with the changes to the import from `ansible_collections....` and as I thought about it, I kind of expected this error, because the collections is not installed on the machine here.  

```
TASK [core/cluster : set | delete node] **********************************************************************
ok: [removenode-x-master]

TASK [core/cluster : Delete an IBM Spectrum Scale Node from Cluster] *****************************************
fatal: [removenode-x-master]: FAILED! => {"msg": "Could not find imported module support code for ibm_ss_node.  Looked for either get_logger.py or ibm_ss_utils.py"}
...
...
```

But what is strange is that the `library` and `module_utils` should be added in locally from the directory of the play.  From this doc: https://docs.ansible.com/ansible/latest/dev_guide/developing_locally.html#adding-a-module-locally

> store it in a sub-directory called library in the directory that contains the playbook(s)

Even adding `try/except` blocks on the import didn't resolve this...    I am able to drop a simple .yml file like this and run it...   getting the same error. 
```
---
- hosts: localhost
  tasks:
     - name: print the cluster
       ibm_ss_cluster:
          op: get
       register: output
```

Will continue to investigate.

whowutwut commented:
@mamuthiah Any idea here?  With the changes on this branch... and a simple playbook:

```
# cat test.yml
---
- hosts: localhost
  tasks:
     - name: print the cluster
       ibm_ss_cluster:
          op: get
       register: output

     - debug: var=output
```

Results in this error: 

```
# ansible-playbook test.yml
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'
...

TASK [print the cluster] *************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"msg": "Could not find imported module support code for ibm_ss_cluster.  Looked for either runCmd.py or ibm_ss_utils.py"}

PLAY RECAP ***************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0

```

This makes sense since it's trying to do `from ansible_collections.ibm_spectrum_scale.install_infra.plugins.module_utils.ibm_ss_utils import runCmd, parse_aggregate_cmd_output, RC_SUCCESS` but this is not installed via collections on my machine. 

If i put the code back `from ansible.module_utils.ibm_ss_utils import runCmd, parse_aggregate_cmd_output, RC_SUCCESS` it works fine...

```
TASK [debug] *************************************************************************************************************************************************************************************
ok: [localhost] => {
    "output": {
        "changed": false,
        "failed": false,
        "msg": "mmlscluster successfully executed",
        "rc": 0,
...
```

But adding a try/except block does NOT  work..

```python
try:
    from ansible_collections.ibm_spectrum_scale.install_infra.plugins.module_utils.ibm_ss_utils import runCmd, parse_aggregate_cmd_output, RC_SUCCESS
except ImportError as ie:
    from ansible.module_utils.ibm_ss_utils import runCmd, parse_aggregate_cmd_output, RC_SUCCESS
```

Fails with similar message: 

```
# ansible-playbook test.yml
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [localhost] *********************************************************************************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************************************************************************************
ok: [localhost]

TASK [print the cluster] *************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"msg": "Could not find imported module support code for ibm_ss_cluster.  Looked for either runCmd.py or ibm_ss_utils.py"}

PLAY RECAP ***************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0

```

Any  thoughts?

whowutwut commented:
Ok, some updates, I think this actually has to do with python2 and python3...  Its not the `try/except` code, I don't think... I created a dev environment following this: https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_general.html and moved the `ibm_ss_cluster` and `ibm_ss_utils` files over to the `modules` and `module_utils` so that  I can run `python -m` directly against the code..

```
# cat /tmp/op.json
{
    "ANSIBLE_MODULE_ARGS": {
        "op": "get"
    }
}
# python -m ansible.modules.ibm_ss_cluster /tmp/op.json 
```

So the first issue is: 
```
Traceback (most recent call last):
...
  File "/root/ansible/lib/ansible/module_utils/ibm_ss_utils.py", line 159, in parse_aggregate_cmd_output
    values = line.split(":")
AttributeError: 'int' object has no attribute 'split'
```

This seems to be caused by the return of runCmd not being a string so the isistance test fails... and we don't run `spitlines()` at https://github.com/IBM/ibm-spectrum-scale-install-infra/blob/master/roles/custom_module/module_utils/ibm_ss_utils.py#L143

After i force splitlines(), we get more issues.... 
```
  File "/root/ansible/lib/ansible/module_utils/ibm_ss_utils.py", line 46, in decode
    return urllib.unquote(input_string)
AttributeError: module 'urllib' has no attribute 'unquote'
```

FYI @mamuthiah

whowutwut commented:
Alright... so lots of new info..   I opened an issue to the ansible project and it was confirmed that we can't use `try/except` blocks here. 

> The ansiballz construction code does not support try/except blocks around module_utils imports, nor are there plans to support this. As such, you cannot write a module that will work with both collections and without. You will need to maintain 2 copies of the module.

I think the changes in this PR are valid, _if we want to go down the path of collections_ and it's up to us to change the way we call the roles..    In some internal discussion right now, is to adopt the following strategy: 

1. We decide collections is the way to go and we merge this PR. Then...
2. On each release, we upload as part of the release the .tar.gz collections file 
3. Users are asked to install collections via release files (it's up to them where they want to install it, global or local)  If they want to track dev, they can clone and build the collections, then install that ... 
4. To directly run the playbook as we do today,  run it from `ansible_collection/ibm_spectrum_scale/install_infra/`  instead of from the git clone 

Eventually (far out) this may be pushed to galaxy...   but until then, cloning of the repo would be reserved for those users who want the latest/greatest dev code, or want to contribute back to source...  consumers would just take the .tar.gz file to get the code onto their machines. 

@troppens you seem to have pointed out galaxy in the past in some issues, do you have any thoughts on this?

troppens commented:
@whowutwut No. I do not have enough experience to comment.

troppens commented:
I did some playbooks to build test & demo environments. I have not seen any Ansible role or module that does nested structures such as `core/cluster`. Initially I tried something like this on my own, but noticed that default variables, etc. does not work nice with nested structures. Therefore I would recommend to avoid this.

whowutwut commented:
@troppens , you do bring up a valid point about nested structures, i have not seen many examples of this. 

But I haven't hit any issues with defaults not being picked up, or not able to override them.  Under the roles, we have tried to init with galaxy, so the directory structure is more standardized. 

Ansible is very loosie-goosie, and it seems like there are "best practices" and something like `ansible-galaxy init role` would set up some initial directory structure... but there's no _requirement_ to follow it, i don't think.    Seems like as long as roles are in the right path, or ansible.cfg is set up to override defaults, it would work, but it's very muddy.

whowutwut commented:
I'm going to close this issue, it's replaced by #172, which is a PR created off of our git subtree changes..  rather than @dunnevan 's forked branch...
